apollo-router/src/plugins/cache/entity.rs

```diff
@@ -1344,22 +1410,55 @@ fn extract_cache_keys(
         // - query hash: invalidate the entry for a specific query and operation name
         // - additional data: separate cache entries depending on info like authorization status
         let mut key = String::new();
-        let _ = write!(&mut key, "version:{ENTITY_CACHE_VERSION}:subgraph:{subgraph_name}:type:{typename}:entity:{hashed_entity_key}:hash:{query_hash}:data:{additional_data_hash}");
+        let _ = write!(&mut key, "version:{ENTITY_CACHE_VERSION}:subgraph:{subgraph_name}:type:{typename}:entity:{hashed_entity_key}:representation:{hashed_representation}:hash:{query_hash}:data:{additional_data_hash}");
         if is_known_private {
             if let Some(id) = private_id {
                 let _ = write!(&mut key, ":{id}");
             }
         }
 
-        representation
-            .as_object_mut()
-            .map(|o| o.insert(TYPENAME, opt_type));
+        representation.insert(TYPENAME, typename_value);
+        representation_entity_keys
+            .into_iter()
+            .for_each(|(key, val)| {
+                representation.insert(key, val);
+            });
         res.push(key);
     }
     Ok(res)
 }
 
-pub(crate) fn hash_entity_key(representation: &Value) -> String {
+/// Use it specifically for keys field set because we don't check for fragment spread
+fn get_root_field_names(selection_set: &SelectionSet) -> Vec<Name> {
```

Comment:
With apollo-compiler 1.27.0 released today `selection_set.root_fields()` returns an iterator that can replace this function